### ISSUE TITLE
cant play video on android client

### ISSUE THREAD

**Issue by antxxxx** [cdhowie/netflix-no-ipv6-dns-proxy#2]
I have this up and running on my network and now see UK content on netfilx from android client and on windows 7 via webpage (which is great as all of my devices connected to a TV are ipv4 only so actually getting this on android devices is now better) 
However, when I try to play a video on android client, I get the 'you are using a proxy' error and cant view it
It works fine on windows 7 via a browser, and netfilx on ipv4 devices work fine, and I can cast from android device to play it on an ipv4 device

**Comment by cdhowie** [comment]
Can you get a log of the DNS queries (server log or pcap/Wireshark dump) between your DNS server and the failing Android device, _after_ rebooting your Android device?  Perhaps there is another domain that needs to be blacklisted.

**Comment by lbdroid** [comment]
I have noticed recently on Android/Chromecast, that the devices are making connections to IP addresses registered to Netflix (AS2906), without ANY corresponding DNS lookup.

In other words, use of DNS/proxy hacks is being directly targeted by the elimination of DNS from the process after a certain point. This is a fairly simple thing to implement on their end, since it just means throwing a redirect-to-IP into an HTTP response.

The solution is much more elegant than DNS hacks, and involves selective routing. Please see https://github.com/ab77/netflix-proxy/issues/180#issuecomment-224166147 for instructions on setting up your router.

While there is still a DNS element in the process, it doesn't interfere at all with the responses; rather it just adds the results of DNS queries matching a pattern to an ipset. The reason for the hybrid approach, is because *.netflix.com is served by AMAZON, rather than IP addresses that Netflix actually owns.

The next thing that Netflix will probably do to work around this, is to add some direct-to-ip redirects to AMAZON IP's. In which case, we just route all of amazon through the tunnel, for any sources that are accessing netflix.

**Comment by cdhowie** [comment]
> I have noticed recently on Android/Chromecast, that the devices are making connections to IP addresses registered to Netflix (AS2906), without ANY corresponding DNS lookup.

Well that's dumb, and doesn't really even solve anything for them.  It's not like we can't mangle packets to change the destination...

I really can't understand the thought process among their engineering group that led to this "solution."  It doesn't make any sense.

**Comment by lbdroid** [comment]
How many regular home TV viewers are capable of that though? A regular home TV viewer might barely be able to change their DNS server address, but selectively routing traffic through a VPN based on an ipset is a lot more involved.

I really doubt that they are overly concerned about "us", just the thousands of people who pay for services like u*nblock.u$ to give them a DNS hack that is easy to set up in their router.

Here is the thing; we KNOW that in the end, there is absolutely NOTHING that they can do to stop us, EXCEPT for restricting access based on the billing address. But even then, my Canadian credit card will accept a US address I gave them, as a billing address. So I'm unstoppable.

**Comment by cdhowie** [comment]
Sure, but if you can set up a DNS proxy then you can probably figure out netfilter and ipset, too.  Your average home user isn't going to know how to do either.

Regardless, all we're trying to do here is force Netflix over the IPv4 Internet since they blocked HE tunnels, and those of us with HE tunnels would like to keep our IPv6 access without losing Netflix access.  We're not trying to circumvent any block to gain access to content Netflix doesn't think we should have access to.  We're just trying to gain access to the same content we've always had available over native IPv4.  They're making this exceedingly difficult for those of us who aren't even trying to bypass geo content restrictions.

It's beyond me why Netflix doesn't try using IPv4 only if it thinks that the IPv6 endpoint is an "unblocker."

**Comment by lbdroid** [comment]
They are NOT worried about anyone who CAN setup a DNS proxy. They're worried about people who pay someone ELSE to set it up for them, since those people represent way over 99% of those who are geounblocking. If you CAN setup a DNS proxy, then it is technically impossible for them to stop you.

I think that I'm understanding what you are doing better with that response... your ISP doesn't have IPv6 support, so you are using a broker to gain access, but _not_ use that tunnel for netflix -- instead connect directly via the IPv4 address from your ISP.

So what you could probably do, is keep an IPv6 ipset of AS2906 and block them with iptables.

**Comment by cdhowie** [comment]
> They are NOT worried about anyone who CAN setup a DNS proxy. They're worried about people who pay someone ELSE to set it up for them, since those people represent way over 99% of those who are geounblocking.

Right, okay.  We're on the same page now.

> your ISP doesn't have IPv6 support, so you are using a broker to gain access, but not use that tunnel for netflix -- instead connect directly via the IPv4 address from your ISP.

Correct.  The HE endpoint is also in the USA, so I have (well... had) access to the same content over both v4 and v6.  I'm just trying to keep both IPv6 access and Netflix.

> So what you could probably do, is keep an IPv6 ipset of AS2906 and block them with iptables.

I may wind up doing this.  So far the only device I'm having trouble with is my Chromecast.  All of my other devices work just fine with the DNS proxy fix.

**Comment by lbdroid** [comment]
They seem to be adding clients to the direct-to-ip blocking approach. Initially, it was impacting me only with Android client, but working fine with Chromecast (even casting from the Android client). 8 days ago, they added Chromecast.

**Comment by cdhowie** [comment]
So here's what I wound up doing:

```
#!/bin/sh

ipset flush netflix-v6 2>/dev/null || ipset create netflix-v6 hash:net family inet6

for range in $(whois -h whois.radb.net -- '-i origin AS2906' | grep ^route6: | awk '{print $NF}'); do
    ipset add netflix-v6 $range
done
```

Then in ip6tables I have the following rule:

```
-A FORWARD -m set --match-set netflix-v6 dst -j REJECT --reject-with icmp6-no-route
```

I'm still receiving the proxy/unblocker error on my Chromecast, and this rule's packet counter is zero.  :/  I will have to tcpdump later tonight and see what exactly is going wrong.

**Comment by lbdroid** [comment]
I wonder if it is connecting to any AWS ipv6 addresses? It shouldn't though, since you're blocking DNS resolution for *.netflix.com.

Make sure that you aren't allowing ANY outbound dns connections that don't come from your dnsmasq instance. For example, if its bypassing your DNS and pulling a AAAA record from 8.8.8.8 or 208.67.222.222 or something.

**Comment by cdhowie** [comment]
Yeah, that though did occur to me a few days ago (hard-coding a different DNS server).  I wonder how they would deal with networks that don't allow outbound DNS requests except via the network's own DNS server?

Hopefully we'll get some answers later when I can reboot my Chromecast to flush the DNS cache and run tcpdump while trying Netflix again.  (I'm not at home right now.)

**Comment by cdhowie** [comment]
I threw some log rules in iptables just to see what might happen.

```
SRC=192.168.3.58 DST=8.8.8.8 ... PROTO=UDP SPT=57425 DPT=53 LEN=40
```

Sigh... okay.  I don't know if Netflix is running on the Chromecast right now, but at least now we have verified that the Chromecast is indeed sending DNS requests past my local DNS server.  Fun fact: I saw the same log entries with source IPs of Android devices.

Since 8.8.8.8 and 8.8.4.4 are anycast, I guess I'll just add them as additional addresses on my LAN interface so my local DNS server will respond for them.

**Comment by lbdroid** [comment]
How about taking everything with DPT=53 and rewrite its DST to your local DNS?

**Comment by cdhowie** [comment]
I don't necessarily want to go that route just yet.  If it becomes necessary I will.  For now, adding 8.8.8.8 and 8.8.4.4 to my LAN-facing interface has solved the problem and the Chromecast is working again.  The ipset rule in ip6tables still hasn't matched anything, so I'm not 100% sure that it's necessary yet -- it doesn't seem that Chromecast or Android Netflix apps are using hard-coded IP addresses for Netflix endpoints, it just seems that they are using hard-coded DNS servers (8.8.8.8) instead of using the ones DHCP provides.  At least that's all I've been able to observe so far.

**Comment by cdhowie** [comment]
@antxxxx Based on this thread, it seems the solution is to add the IP addresses 8.8.8.8 and 8.8.4.4 to your router's LAN-facing interface.  This assumes the router is running the DNS proxy -- if it's not, you'll have to have your router NAT requests to 8.8.8.8/8.8.4.4 to the host running the proxy.  Can you try this and confirm if it works?

**Comment by antxxxx** [comment]
i have done the following and now UK netflix works from my android device

create alias for 8.8.8.8 and 8.8.4.4 on my host running the proxy and dnsmasq
change resolv.conf on this host to use my ISPs DNS servers rather than 8.8.8.8 and 8.8.4.4
Ensure proxy is listening on 8.8.8.8 and 8.8.4.4
create static route on my router to send 8.8.8.8 and 8.8.4.4 to the host running the proxy

Netflix now displays and plays on my android device

Although I do really appreciate the work you have done on writing this and helping me get it working, I do wish it was not necessary. My ISP does not provide IPv6 and so i use he.net to get IPv6. I understand Netflixs view that using he.net allows easy bypass of geo restriction (though I dont agree with it), but I wish they could just provide an option in clients to disable IPv6, or some other method of region locking on devices

**Comment by cdhowie** [comment]
Thanks for confirming.  This is "fixed" by df29b85.

> Although I do really appreciate the work you have done on writing this and helping me get it working, I do wish it was not necessary. My ISP does not provide IPv6 and so i use he.net to get IPv6.

Trust me, we are in the same boat.  I'm in the exact same situation with respect to IPv6 and I've sunk hours into this just to make Netflix work again.  At least I've learned some things in the process, so it's not all wasted effort...

**Comment by jamesmacwhite** [comment]
I know this is closed, but I came across a similar issue when casting via the Netflix app. I got the dreaded proxy/VPN message of doom after implementing this awesome DNS proxy. I wasn't aware the Chromecast DNS is hard coded and figured DNS settings pushed via DHCP would be OK. The only difference in my case is I'm using conditional forwarders for netflix.com already. So only netflix.com requests actually go through this proxy in the first place. I've deployed it on a multiple Linux machines for a little bit of redundancy.

It appears that the Chromecast device itself will fallback to the set DNS via DHCP, if the Google DNS servers cannot be reached:

With this in mind, I've blocked them at the router level with `iptables`:

```
iptables -I FORWARD --destination 8.8.8.8 -j REJECT
iptables -I FORWARD --destination 8.8.4.4 -j REJECT
```

You can then ping both DNS servers to confirm that the block is in place:

```
Pinging 8.8.8.8 with 32 bytes of data:
Reply from 192.168.1.1: Destination port unreachable.
Reply from 192.168.1.1: Destination port unreachable.
Reply from 192.168.1.1: Destination port unreachable.
Reply from 192.168.1.1: Destination port unreachable.
```

```
Pinging 8.8.4.4 with 32 bytes of data:
Reply from 192.168.1.1: Destination port unreachable.
Reply from 192.168.1.1: Destination port unreachable.
Reply from 192.168.1.1: Destination port unreachable.
Reply from 192.168.1.1: Destination port unreachable.
```

Restart the Chromecast device, and it should now honour the DNS settings being pushed via DHCP, meaning the Netflix DNS Proxy magic can now work on Chromecast too! You shouldn't need to redirect the DNS requests themselves with DNAT, you just want the Chromecast to actually use the DNS settings configured at router level.

Hope this helps.
